src/main/scala/edu/twitter/model/impl/neuralnetwork/DataIterator.scala

```diff
@@ -0,0 +1,179 @@
+  final private val tokenizerFactory: TokenizerFactory = new DefaultTokenizerFactory
+  tokenizerFactory.setTokenPreProcessor(new CommonPreprocessor)
+
+  private var _cursor: Int = 0
+
+  def next(num: Int): DataSet = {
+    if (_cursor >= dataList.length)
+      throw new NoSuchElementException
+    try {
+      nextDataSet(num)
+    } catch {
+      case e: IOException =>
+        throw new RuntimeException(e)
+    }
+  }
+
+  @throws[IOException]
+  private def nextDataSet(num: Int): DataSet = {
+    val reviews = new util.ArrayList[String]()
+    val positive = new util.ArrayList[Boolean]()
+    var i = 0
+
+    while (i < num && _cursor < totalExamples) {
+
+      val msg = dataList(_cursor).getAs[String]("msg")
+      val label = dataList(_cursor).getAs[Double]("label")
+
+      //TODO(elhawaty): Find why some rows return null.
+      if (msg != null) {
+        reviews.add(msg)
+        positive.add(if (label == 1.0) true else false)
+        i += 1
+      } else {
+        println("Invalid tweet:" + dataList(_cursor))
+      }
+
+      _cursor += 1
+    }
+
+    //Second: tokenize reviews and filter out unknown words
+    val allTokens: util.List[util.List[String]] = new util.ArrayList[util.List[String]](reviews.size)
+    var maxLength: Int = 0
+    import scala.collection.JavaConversions._
+    for (s <- reviews) {
+      val tokens = tokenizerFactory.create(s).getTokens
+      val tokensFiltered = new util.ArrayList[String]
+      for (t <- tokens.asScala) {
+        if (wordVectors.hasWord(t)) tokensFiltered.add(t)
+      }
+      allTokens.add(tokensFiltered)
+      maxLength = Math.max(maxLength, tokensFiltered.size)
+    }
+
+    // Workaround, just in case the word2vec doesn't recognise all the words in the batch which is unlikely to happen as we are using Google word2vec.
+    if (maxLength == 0) {
+      println(reviews.get(0))
+      allTokens.get(0).add("times")
+      maxLength = 1;
+    }
+
+    //If longest review exceeds 'truncateLength': only take the first 'truncateLength' words
+    if (maxLength > truncateLength) maxLength = truncateLength
+
+    //Create data for training
+    //Here: we have reviews.size() examples of varying lengths
+    val features = Nd4j.create(reviews.size, vectorSize, maxLength)
+    val labels = Nd4j.create(reviews.size, 2, maxLength)
+    //Two labels: positive or negative because we are dealing with reviews of different lengths and only one output at the final time step: use padding arrays mask arrays contain 1 if data is present at that time step for that example, or 0 if data is just padding
+    val featuresMask = Nd4j.zeros(reviews.size, maxLength)
+    val labelsMask = Nd4j.zeros(reviews.size, maxLength)
+
+    val temp = new Array[Int](2)
+    for (i <- reviews.indices) {
+      val tokens: util.List[String] = allTokens.get(i)
+      temp(0) = i
+      //Get word vectors for each word in review, and put them in the training data
+      var j: Int = 0
```

Comment:
no need for this j? or u need this j and for loop in next j to be something else?

Reply:
Done 👍